### DISCUSSION TITLE
ZTransducer Vs ZPipeline

### DISCUSSION THREAD

**Discussion by AAAboutajedyne** [zio/zio#6262] [Q&A] [answered]
Since ZTransducer can do what sink could do, but faster to run comparing to sinks (Transducer concept). What is the purpose of adding ZPipeline that can do what ZTransducer could do. Knowing that with ZTransducer we can abstract over stream transformations at the level of values, creating, storing, and passing around reusable transformation that can be applied to many different streams ? Can anyone help me see the need of ZPipeline ?

**Comment by adamgfraser** [comment, marked as answer]
`ZPipeline` basically is the new `ZTransducer`. It does exactly that, describing the "middle" of transformations that go from stream to stream whereas streams are the "beginning" and sinks are the "end". Pipelines can describe transduction operations but are also more general in that pipelines are channels themselves. Note that `ZTransducer` doesn't exist in ZIO 2.0 so you're right that we don't need both.

**Reply by AAAboutajedyne** [reply]
Ok, thank you. I was checking https://javadoc.io/static/dev.zio/zio-streams_2.13/2.0.0-M4/zio/stream/ZTransducer$.html and ZTransducer are not deprecated. It will be in next releases ?

**Reply by vigoo** [reply]
`2.0.0-M4` still contained the old streams implementation, together with the new one in the `zio.stream.experimental` package. The current latest, `2.0.0-RC1` only has the new one (in `zio.streams`)

**Reply by AAAboutajedyne** [reply]
Nice, makes sense :)
